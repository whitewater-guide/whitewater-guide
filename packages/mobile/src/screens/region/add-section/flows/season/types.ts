import Animated from 'react-native-reanimated';

export interface AnimatedPoint {
  x: Animated.Value<number>;
  y: Animated.Value<number>;
}
