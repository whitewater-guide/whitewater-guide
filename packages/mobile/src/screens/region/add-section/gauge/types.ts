import { Gauge } from '@whitewater-guide/commons';

export type ListedGauge = Pick<
  Gauge,
  'id' | 'name' | 'code' | 'levelUnit' | 'flowUnit' | 'lastMeasurement'
>;
