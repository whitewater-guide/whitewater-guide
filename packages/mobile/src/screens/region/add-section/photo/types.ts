import { CompositeNavigationProp, RouteProp } from '@react-navigation/native';
import { StackNavigationProp } from '@react-navigation/stack';
import { Screens } from '~/core/navigation';
import {
  AddSectionNavProp,
  AddSectionStackParamsList,
} from '~/screens/region/add-section/types';

export type AddSectionPhotoNavProp = CompositeNavigationProp<
  StackNavigationProp<AddSectionStackParamsList, Screens.ADD_SECTION_PHOTO>,
  AddSectionNavProp
>;

export interface AddSectionPhotoNavProps {
  navigation: AddSectionNavProp;
  route: RouteProp<AddSectionStackParamsList, Screens.ADD_SECTION_PHOTO>;
}
