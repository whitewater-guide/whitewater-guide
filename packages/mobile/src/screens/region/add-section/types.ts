import { Coordinate3d } from '@whitewater-guide/commons';

export type Uncoordinate = [
  string | undefined,
  string | undefined,
  string | undefined,
];

export interface Shape {
  shape: [Uncoordinate, Uncoordinate];
}

export type Maybe3d = Coordinate3d | undefined;
