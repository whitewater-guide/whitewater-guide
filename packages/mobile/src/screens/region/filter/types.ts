import { CompositeNavigationProp } from '@react-navigation/native';
import { StackNavigationProp } from '@react-navigation/stack';
import { SectionFilterOptions } from '@whitewater-guide/clients';

import { Screens } from '~/core/navigation';
import { SelectableTag } from '~/features/tags';
import {
  RegionScreenNavProp,
  RegionStackParamsList,
} from '~/screens/region/types';

export interface SearchState
  extends Omit<SectionFilterOptions, 'withTags' | 'withoutTags'> {
  kayaking: SelectableTag[];
  hazards: SelectableTag[];
  supply: SelectableTag[];
  misc: SelectableTag[];
}

export type FilterNavProp = CompositeNavigationProp<
  StackNavigationProp<RegionStackParamsList, Screens.FILTER>,
  RegionScreenNavProp
>;

export interface FilterNavProps {
  navigation: FilterNavProp;
}
