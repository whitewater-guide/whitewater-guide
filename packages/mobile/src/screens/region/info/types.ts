import type { MaterialBottomTabNavigationProp } from '@react-navigation/material-bottom-tabs';
import type {
  CompositeNavigationProp,
  RouteProp,
} from '@react-navigation/native';

import type { Screens } from '~/core/navigation';
import type {
  RegionTabsNavProp,
  RegionTabsParamsList,
} from '~/screens/region/types';

export type RegionInfoNavProp = CompositeNavigationProp<
  MaterialBottomTabNavigationProp<RegionTabsParamsList, Screens.REGION_INFO>,
  RegionTabsNavProp
>;

export interface RegionInfoNavProps {
  navigation: RegionInfoNavProp;
  route: RouteProp<RegionTabsParamsList, Screens.REGION_INFO>;
}
