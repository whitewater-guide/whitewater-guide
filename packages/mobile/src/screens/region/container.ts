import { graphql } from 'react-apollo';
import { connect } from 'react-redux';
import { compose, mapProps, pure } from 'recompose';
import { withLoading } from '../../components';
import { chunkedListLoader, queryResultToNode } from '../../ww-clients/apollo';
import { withFeatureIds } from '../../ww-clients/core';
import { REGION_DETAILS, WithRegion } from '../../ww-clients/features/regions';
import { searchTermsSelector, WithSearchTerms } from '../../ww-clients/features/regions/selectors';
import { WithSectionsList, withSectionsList } from '../../ww-clients/features/sections';
import { applySearch } from '../../ww-commons';

export default compose(
  withFeatureIds('region'),
  graphql(
    REGION_DETAILS,
    {
      alias: 'withSection',
      options: () => ({ fetchPolicy: 'cache-and-network' }),
      props: (props) => queryResultToNode(props, 'region'),
    },
  ),
  withLoading<WithRegion>((props) => props.region.loading),
  withSectionsList(),
  chunkedListLoader('sections'),
  connect<WithSearchTerms, {}, WithSectionsList>(searchTermsSelector),
  mapProps(({ sections, searchTerms, ...props }) => ({
    ...props,
    searchTerms,
    sections: {
      ...sections,
      nodes: applySearch(sections.nodes, searchTerms),
    },
  })),
  pure,
);
