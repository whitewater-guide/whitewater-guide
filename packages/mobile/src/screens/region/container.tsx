import { ApolloQueryResult, ObservableQuery } from 'apollo-client';
import React from 'react';
import { withApollo, WithApolloClient } from 'react-apollo';
import { withNetworkConnectivity } from 'react-native-offline';
import { connect } from 'react-redux';
import { compose } from 'recompose';
import { RootState } from '../../core/reducers';
import { getListMerger } from '../../ww-clients/apollo';
import { LIST_SECTIONS, Result, Vars } from '../../ww-clients/features/sections';
import { applySearch, Section } from '../../ww-commons';
import { InnerProps, OuterProps, SectionsStatus } from './types';

interface ConnectivityProps {
  isConnected: boolean;
}

interface State {
  sections: Section[];
  count: number;
  status: SectionsStatus;
}

type WrapperProps<P> = OuterProps & ConnectivityProps & WithApolloClient<P>;

function container(limit: number = 20) {
  // tslint:disable-next-line:only-arrow-functions
  return function<P>(Wrapped: React.ComponentType<P & InnerProps>): React.ComponentType<P & OuterProps> {
    class WithRegionSections extends React.PureComponent<WrapperProps<P>, State> {
      _query!: ObservableQuery<Result, Vars>;
      _subscription!: ZenObservable.Subscription;

      constructor(props: WrapperProps<P>) {
        super(props);
        const { client, region  } = props;
        if (!region.node) {
          return;
        }

        let fromCache: Result | null = null;
        try {
          fromCache = client.readQuery<Result, Vars>({
            query: LIST_SECTIONS,
            variables: { filter: { regionId: region.node.id } },
          });
        } catch (e) {/* ignore, not in cache*/}
        this.state = {
          sections: fromCache ? fromCache.sections.nodes! : [],
          count: fromCache ? fromCache.sections.count! : 0,
          status: SectionsStatus.READY,
        };

        this._query = client.watchQuery<Result, Vars>({
          query: LIST_SECTIONS,
          fetchPolicy: 'cache-only',
          variables: { filter: { regionId: region.node.id } },
        });
        this._subscription = this._query.subscribe(this.onUpdate);
      }

      async componentDidMount() {
        const { count, sections } = this.state;
        if (count === 0 || sections.length < count) {
          await this.loadInitial();
        } else {
          await this.loadUpdates();
        }
        // TODO: start polling
      }

      componentWillUnmount() {
        this._subscription.unsubscribe();
      }

      onUpdate = ({ data }: ApolloQueryResult<Result>) => {
        // this will happen when trying to read from cache when cache is empty
        if (!data.sections) {
          return;
        }
        this.setState({ sections: data.sections.nodes!, count: data.sections.count! });
      };

      /**
       * This will load sections in chunks
       * @param offset
       * @param updatedAfter
       */
      loadSections = async (offset: number = 0, updatedAfter?: Date) => {
        const { isConnected, region } = this.props;
        if (!region.node) {
          return;
        }
        if (!isConnected) {
          return;
        }

        const { data } = await this._query.fetchMore({
          query: LIST_SECTIONS,
          variables: {  page: { limit, offset }, filter: { regionId: region.node.id, updatedAfter } },
          updateQuery: getListMerger('sections') as any,
        });

        const { sections: { nodes, count } } = data;
        if (offset + nodes!.length < count!) {
          await this.loadSections(offset + nodes!.length, updatedAfter);
        }
      };

      loadInitial = async () => {
        this.setState({ status: SectionsStatus.LOADING });
        await this.loadSections(this.state.sections.length);
        this.setState({ status: SectionsStatus.READY });
      };

      loadUpdates = async () => {
        // get latest updated section
        const updatedAfter: Date | undefined = this.state.sections.reduce(
          (acc, { updatedAt }) => (acc && acc > new Date(updatedAt)) ? acc : new Date(updatedAt),
          undefined as (Date | undefined),
        );
        this.setState({ status: SectionsStatus.LOADING_UPDATES });
        this.loadSections(0, updatedAfter);
        this.setState({ status: SectionsStatus.READY });
      };

      refresh = () => {
        const { sections, count, status } = this.state;
        if (status !== SectionsStatus.READY) {
          return;
        }
        if (sections.length < count) {
          this.loadInitial();
        } else {
          this.loadUpdates();
        }
      };

      render() {
        const { searchTerms, isConnected, client, ...props } = this.props as any;
        const { status, count, sections } = this.state;
        const filteredSections = applySearch(sections, searchTerms);
        return (
          <Wrapped
            {...props}
            status={status}
            count={count}
            sections={filteredSections}
            refresh={this.refresh}
          />
        );
      }
    }

    return compose<WrapperProps<P>, OuterProps & P>(
      connect((state: RootState) => ({ isConnected: state.network.isConnected })),
      withApollo,
    )(WithRegionSections);
  };
}

export default container;
