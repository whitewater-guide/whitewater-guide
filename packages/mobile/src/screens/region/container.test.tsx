import { mount } from 'enzyme';
import { MockList } from 'graphql-tools';
import React from 'react';
import { Provider } from 'react-redux';
import createMockStore from 'redux-mock-store';
import { Overwrite } from 'type-zoo';
import { LIST_SECTIONS, Result, Vars } from '../../ww-clients/features/sections';
import {
  createFixedProvider,
  createMockedProvider,
  findReceiver,
  FixedProviderOptions,
  flushPromises,
  genericReceiver,
  Receiver,
} from '../../ww-clients/test';
import { Section, SectionSearchTerms } from '../../ww-commons';
import container from './container';
import { InnerProps, SectionsStatus } from './types';

let seedSections: Section[];
let seedCount: number;
let seedData: Result;

beforeAll(async () => {
  const provider = createMockedProvider({
    mocks: {
      SectionsList: () => ({
        nodes: () => new MockList(5),
        count: () => 5,
      }),
    },
  });
  const result = await provider.client.query<Result, Vars>({
    query: LIST_SECTIONS,
  });
  seedData = result.data;
  seedSections = result.data.sections.nodes!;
  seedCount = result.data.sections.count!;
});

interface TestOptions extends Overwrite<FixedProviderOptions<Result, Vars>, {cache?: Result}> {
  isConnected?: boolean;
  searchTerms?: SectionSearchTerms;
}

const mountInHarness = async (options: TestOptions): Promise<Receiver<InnerProps>> => {
  const { isConnected = true, searchTerms = null, cache, ...opts } = options;
  const store = createMockStore()({ network: { isConnected } });
  const FixedProvider = createFixedProvider({
    ...opts,
    cache: cache && { query: LIST_SECTIONS, data: cache, variables: { filter: { regionId: 'test_region_id' } } },
  });
  const WithData = container()(genericReceiver()) as any;
  const wrapped = mount(
    <Provider store={store}>
      <FixedProvider>
        <WithData region={{ node: { id: 'test_region_id' } }} searchTerms={searchTerms} />
      </FixedProvider>
    </Provider>,
  );
  const receiver = findReceiver(wrapped);
  await flushPromises();
  return receiver;
};

it('seed data should match schema', () => {
  expect(seedData).toMatchSnapshot();
});

it('should read result from cache first', async () => {
  const receiver = await mountInHarness({
    cache: seedData,
    responses: [],
  });

  expect(receiver.props.count).toEqual(seedCount);
  expect(receiver.props.status).toEqual(SectionsStatus.READY);
  expect(receiver.props.sections).toEqual(seedSections);
});

it('should pass empty list first when cache is empty', async () => {
  const receiver = await mountInHarness({
    responses: [],
  });

  expect(receiver.cdm.count).toEqual(0);
  expect(receiver.cdm.status).toEqual(SectionsStatus.READY);
  expect(receiver.cdm.sections).toEqual([]);
});

it('should load initial data when cache contains no data', async () => {

});

it('should continue loading initial data when cache contains partial data', async () => {

});

it('should paginate initial data', async () => {

});

it('should set state to ready after all initial data has been loaded', async () => {

});

it('should start loading updates when cache is full', async () => {

});

it('should paginate updates', async () => {

});

it('should update sections count if updates contain new sections', async () => {

});

it('should set state to ready after all updates are loaded', async () => {

});

it('should not try to load anything when offline', async () => {

});

it('should release subscription when unmounted', async () => {

});

it('should poll lastMeasurements', async () => {

});

it('should pass updates when is changed outside the query', () => {

});

it('should apply filters', async () => {

});

it('something about refresh', async () => {

});
