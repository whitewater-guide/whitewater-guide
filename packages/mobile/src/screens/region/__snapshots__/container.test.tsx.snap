// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`seed data should match schema 1`] = `
Object {
  "sections": Object {
    "__typename": "SectionsList",
    "count": 5,
    "nodes": Array [
      Object {
        "__typename": "Section",
        "altNames": Array [
          "Section.altNames.1",
          "Section.altNames.2",
        ],
        "demo": false,
        "difficulty": 1.5,
        "difficultyXtra": "Section.difficultyXtra.1",
        "distance": 1.5,
        "drop": 1.5,
        "duration": 1,
        "flows": Object {
          "__typename": "GaugeBinding",
          "approximate": true,
          "impossible": 2.5,
          "maximum": 2.5,
          "minimum": 2.5,
          "optimum": 2.5,
        },
        "flowsText": "Section.flowsText.1",
        "gauge": Object {
          "__typename": "Gauge",
          "code": "Gauge.code.1",
          "flowUnit": "Gauge.flowUnit.1",
          "id": "Gauge.id.1",
          "lastMeasurement": Object {
            "__typename": "Measurement",
            "flow": 1.5,
            "level": 1.5,
            "timestamp": 2017-02-01T21:00:00.000Z,
          },
          "levelUnit": "Gauge.levelUnit.1",
          "location": Object {
            "__typename": "Point",
            "coordinates": Array [
              5.5,
              6.5,
            ],
            "id": "Point.id.3",
            "kind": "Point.kind.3",
          },
          "name": "Gauge.name.1",
          "source": Object {
            "__typename": "Source",
            "id": "Source.id.1",
            "name": "Source.name.1",
            "termsOfUse": "Source.termsOfUse.1",
          },
          "url": "Gauge.url.1",
        },
        "hidden": false,
        "id": "Section.id.1",
        "levels": Object {
          "__typename": "GaugeBinding",
          "approximate": false,
          "impossible": 1.5,
          "maximum": 1.5,
          "minimum": 1.5,
          "optimum": 1.5,
        },
        "name": "Section.name.1",
        "putIn": Object {
          "__typename": "Point",
          "coordinates": Array [
            1.5,
            2.5,
          ],
          "id": "Point.id.1",
          "kind": "Point.kind.1",
        },
        "rating": 1.5,
        "river": Object {
          "__typename": "River",
          "altNames": Array [
            "River.altNames.1",
            "River.altNames.2",
          ],
          "id": "River.id.1",
          "name": "River.name.1",
        },
        "season": "Section.season.1",
        "seasonNumeric": Array [
          1,
          2,
        ],
        "shape": Array [
          Array [
            1.5,
            2.5,
          ],
          Array [
            3.5,
            4.5,
          ],
        ],
        "tags": Array [
          Object {
            "__typename": "Tag",
            "category": "hazards",
            "id": "Tag.id.1",
            "name": "Tag.name.1",
          },
          Object {
            "__typename": "Tag",
            "category": "supply",
            "id": "Tag.id.2",
            "name": "Tag.name.2",
          },
        ],
        "takeOut": Object {
          "__typename": "Point",
          "coordinates": Array [
            3.5,
            4.5,
          ],
          "id": "Point.id.2",
          "kind": "Point.kind.2",
        },
        "updatedAt": 2017-02-01T21:00:00.000Z,
      },
      Object {
        "__typename": "Section",
        "altNames": Array [
          "Section.altNames.3",
          "Section.altNames.4",
        ],
        "demo": true,
        "difficulty": 2.5,
        "difficultyXtra": "Section.difficultyXtra.2",
        "distance": 2.5,
        "drop": 2.5,
        "duration": 2,
        "flows": Object {
          "__typename": "GaugeBinding",
          "approximate": true,
          "impossible": 4.5,
          "maximum": 4.5,
          "minimum": 4.5,
          "optimum": 4.5,
        },
        "flowsText": "Section.flowsText.2",
        "gauge": Object {
          "__typename": "Gauge",
          "code": "Gauge.code.2",
          "flowUnit": "Gauge.flowUnit.2",
          "id": "Gauge.id.2",
          "lastMeasurement": Object {
            "__typename": "Measurement",
            "flow": 2.5,
            "level": 2.5,
            "timestamp": 2017-02-02T21:00:00.000Z,
          },
          "levelUnit": "Gauge.levelUnit.2",
          "location": Object {
            "__typename": "Point",
            "coordinates": Array [
              11.5,
              12.5,
            ],
            "id": "Point.id.6",
            "kind": "Point.kind.6",
          },
          "name": "Gauge.name.2",
          "source": Object {
            "__typename": "Source",
            "id": "Source.id.2",
            "name": "Source.name.2",
            "termsOfUse": "Source.termsOfUse.2",
          },
          "url": "Gauge.url.2",
        },
        "hidden": true,
        "id": "Section.id.2",
        "levels": Object {
          "__typename": "GaugeBinding",
          "approximate": false,
          "impossible": 3.5,
          "maximum": 3.5,
          "minimum": 3.5,
          "optimum": 3.5,
        },
        "name": "Section.name.2",
        "putIn": Object {
          "__typename": "Point",
          "coordinates": Array [
            7.5,
            8.5,
          ],
          "id": "Point.id.4",
          "kind": "Point.kind.4",
        },
        "rating": 2.5,
        "river": Object {
          "__typename": "River",
          "altNames": Array [
            "River.altNames.3",
            "River.altNames.4",
          ],
          "id": "River.id.2",
          "name": "River.name.2",
        },
        "season": "Section.season.2",
        "seasonNumeric": Array [
          3,
          4,
        ],
        "shape": Array [
          Array [
            5.5,
            6.5,
          ],
          Array [
            7.5,
            8.5,
          ],
        ],
        "tags": Array [
          Object {
            "__typename": "Tag",
            "category": "misc",
            "id": "Tag.id.3",
            "name": "Tag.name.3",
          },
          Object {
            "__typename": "Tag",
            "category": "kayaking",
            "id": "Tag.id.4",
            "name": "Tag.name.4",
          },
        ],
        "takeOut": Object {
          "__typename": "Point",
          "coordinates": Array [
            9.5,
            10.5,
          ],
          "id": "Point.id.5",
          "kind": "Point.kind.5",
        },
        "updatedAt": 2017-02-02T21:00:00.000Z,
      },
      Object {
        "__typename": "Section",
        "altNames": Array [
          "Section.altNames.5",
          "Section.altNames.6",
        ],
        "demo": false,
        "difficulty": 3.5,
        "difficultyXtra": "Section.difficultyXtra.3",
        "distance": 3.5,
        "drop": 3.5,
        "duration": 3,
        "flows": Object {
          "__typename": "GaugeBinding",
          "approximate": true,
          "impossible": 6.5,
          "maximum": 6.5,
          "minimum": 6.5,
          "optimum": 6.5,
        },
        "flowsText": "Section.flowsText.3",
        "gauge": Object {
          "__typename": "Gauge",
          "code": "Gauge.code.3",
          "flowUnit": "Gauge.flowUnit.3",
          "id": "Gauge.id.3",
          "lastMeasurement": Object {
            "__typename": "Measurement",
            "flow": 3.5,
            "level": 3.5,
            "timestamp": 2017-02-03T21:00:00.000Z,
          },
          "levelUnit": "Gauge.levelUnit.3",
          "location": Object {
            "__typename": "Point",
            "coordinates": Array [
              17.5,
              18.5,
            ],
            "id": "Point.id.9",
            "kind": "Point.kind.9",
          },
          "name": "Gauge.name.3",
          "source": Object {
            "__typename": "Source",
            "id": "Source.id.3",
            "name": "Source.name.3",
            "termsOfUse": "Source.termsOfUse.3",
          },
          "url": "Gauge.url.3",
        },
        "hidden": false,
        "id": "Section.id.3",
        "levels": Object {
          "__typename": "GaugeBinding",
          "approximate": false,
          "impossible": 5.5,
          "maximum": 5.5,
          "minimum": 5.5,
          "optimum": 5.5,
        },
        "name": "Section.name.3",
        "putIn": Object {
          "__typename": "Point",
          "coordinates": Array [
            13.5,
            14.5,
          ],
          "id": "Point.id.7",
          "kind": "Point.kind.7",
        },
        "rating": 3.5,
        "river": Object {
          "__typename": "River",
          "altNames": Array [
            "River.altNames.5",
            "River.altNames.6",
          ],
          "id": "River.id.3",
          "name": "River.name.3",
        },
        "season": "Section.season.3",
        "seasonNumeric": Array [
          5,
          6,
        ],
        "shape": Array [
          Array [
            9.5,
            10.5,
          ],
          Array [
            11.5,
            12.5,
          ],
        ],
        "tags": Array [
          Object {
            "__typename": "Tag",
            "category": "hazards",
            "id": "Tag.id.5",
            "name": "Tag.name.5",
          },
          Object {
            "__typename": "Tag",
            "category": "supply",
            "id": "Tag.id.6",
            "name": "Tag.name.6",
          },
        ],
        "takeOut": Object {
          "__typename": "Point",
          "coordinates": Array [
            15.5,
            16.5,
          ],
          "id": "Point.id.8",
          "kind": "Point.kind.8",
        },
        "updatedAt": 2017-02-03T21:00:00.000Z,
      },
      Object {
        "__typename": "Section",
        "altNames": Array [
          "Section.altNames.7",
          "Section.altNames.8",
        ],
        "demo": true,
        "difficulty": 4.5,
        "difficultyXtra": "Section.difficultyXtra.4",
        "distance": 4.5,
        "drop": 4.5,
        "duration": 4,
        "flows": Object {
          "__typename": "GaugeBinding",
          "approximate": true,
          "impossible": 8.5,
          "maximum": 8.5,
          "minimum": 8.5,
          "optimum": 8.5,
        },
        "flowsText": "Section.flowsText.4",
        "gauge": Object {
          "__typename": "Gauge",
          "code": "Gauge.code.4",
          "flowUnit": "Gauge.flowUnit.4",
          "id": "Gauge.id.4",
          "lastMeasurement": Object {
            "__typename": "Measurement",
            "flow": 4.5,
            "level": 4.5,
            "timestamp": 2017-02-04T21:00:00.000Z,
          },
          "levelUnit": "Gauge.levelUnit.4",
          "location": Object {
            "__typename": "Point",
            "coordinates": Array [
              23.5,
              24.5,
            ],
            "id": "Point.id.12",
            "kind": "Point.kind.12",
          },
          "name": "Gauge.name.4",
          "source": Object {
            "__typename": "Source",
            "id": "Source.id.4",
            "name": "Source.name.4",
            "termsOfUse": "Source.termsOfUse.4",
          },
          "url": "Gauge.url.4",
        },
        "hidden": true,
        "id": "Section.id.4",
        "levels": Object {
          "__typename": "GaugeBinding",
          "approximate": false,
          "impossible": 7.5,
          "maximum": 7.5,
          "minimum": 7.5,
          "optimum": 7.5,
        },
        "name": "Section.name.4",
        "putIn": Object {
          "__typename": "Point",
          "coordinates": Array [
            19.5,
            20.5,
          ],
          "id": "Point.id.10",
          "kind": "Point.kind.10",
        },
        "rating": 4.5,
        "river": Object {
          "__typename": "River",
          "altNames": Array [
            "River.altNames.7",
            "River.altNames.8",
          ],
          "id": "River.id.4",
          "name": "River.name.4",
        },
        "season": "Section.season.4",
        "seasonNumeric": Array [
          7,
          8,
        ],
        "shape": Array [
          Array [
            13.5,
            14.5,
          ],
          Array [
            15.5,
            16.5,
          ],
        ],
        "tags": Array [
          Object {
            "__typename": "Tag",
            "category": "misc",
            "id": "Tag.id.7",
            "name": "Tag.name.7",
          },
          Object {
            "__typename": "Tag",
            "category": "kayaking",
            "id": "Tag.id.8",
            "name": "Tag.name.8",
          },
        ],
        "takeOut": Object {
          "__typename": "Point",
          "coordinates": Array [
            21.5,
            22.5,
          ],
          "id": "Point.id.11",
          "kind": "Point.kind.11",
        },
        "updatedAt": 2017-02-04T21:00:00.000Z,
      },
      Object {
        "__typename": "Section",
        "altNames": Array [
          "Section.altNames.9",
          "Section.altNames.10",
        ],
        "demo": false,
        "difficulty": 5.5,
        "difficultyXtra": "Section.difficultyXtra.5",
        "distance": 5.5,
        "drop": 5.5,
        "duration": 5,
        "flows": Object {
          "__typename": "GaugeBinding",
          "approximate": true,
          "impossible": 10.5,
          "maximum": 10.5,
          "minimum": 10.5,
          "optimum": 10.5,
        },
        "flowsText": "Section.flowsText.5",
        "gauge": Object {
          "__typename": "Gauge",
          "code": "Gauge.code.5",
          "flowUnit": "Gauge.flowUnit.5",
          "id": "Gauge.id.5",
          "lastMeasurement": Object {
            "__typename": "Measurement",
            "flow": 5.5,
            "level": 5.5,
            "timestamp": 2017-02-05T21:00:00.000Z,
          },
          "levelUnit": "Gauge.levelUnit.5",
          "location": Object {
            "__typename": "Point",
            "coordinates": Array [
              29.5,
              30.5,
            ],
            "id": "Point.id.15",
            "kind": "Point.kind.15",
          },
          "name": "Gauge.name.5",
          "source": Object {
            "__typename": "Source",
            "id": "Source.id.5",
            "name": "Source.name.5",
            "termsOfUse": "Source.termsOfUse.5",
          },
          "url": "Gauge.url.5",
        },
        "hidden": false,
        "id": "Section.id.5",
        "levels": Object {
          "__typename": "GaugeBinding",
          "approximate": false,
          "impossible": 9.5,
          "maximum": 9.5,
          "minimum": 9.5,
          "optimum": 9.5,
        },
        "name": "Section.name.5",
        "putIn": Object {
          "__typename": "Point",
          "coordinates": Array [
            25.5,
            26.5,
          ],
          "id": "Point.id.13",
          "kind": "Point.kind.13",
        },
        "rating": 5.5,
        "river": Object {
          "__typename": "River",
          "altNames": Array [
            "River.altNames.9",
            "River.altNames.10",
          ],
          "id": "River.id.5",
          "name": "River.name.5",
        },
        "season": "Section.season.5",
        "seasonNumeric": Array [
          9,
          10,
        ],
        "shape": Array [
          Array [
            17.5,
            18.5,
          ],
          Array [
            19.5,
            20.5,
          ],
        ],
        "tags": Array [
          Object {
            "__typename": "Tag",
            "category": "hazards",
            "id": "Tag.id.9",
            "name": "Tag.name.9",
          },
          Object {
            "__typename": "Tag",
            "category": "supply",
            "id": "Tag.id.10",
            "name": "Tag.name.10",
          },
        ],
        "takeOut": Object {
          "__typename": "Point",
          "coordinates": Array [
            27.5,
            28.5,
          ],
          "id": "Point.id.14",
          "kind": "Point.kind.14",
        },
        "updatedAt": 2017-02-05T21:00:00.000Z,
      },
    ],
  },
}
`;
